niworkflows/anat/mni.py

```diff
@@ -214,14 +342,102 @@ def _validate_results(self):
+    """
+    Applies a binary mask to an image.
+
+    Parameters
+    ----------
+    in_file : str
+              Path to a NIfTI file.
+    mask_file : str
+                Path to a NIfTI file.
+    new_name : str
+               Path/filename for the masked output image.
+
+    Returns
+    -------
+    str
+        Absolute path of the masked output image.
+
+    Notes
+    -----
+    in_file and mask_file must be in the same
+    image space and have the same dimensions.
+    """
     import nibabel as nb
     import os
-
+    # Load the input image
     in_nii = nb.load(in_file)
+    # Load the mask image
     mask_nii = nb.load(mask_file)
+    # Set all non-mask voxels in the input file to zero.
     data = in_nii.get_data()
     data[mask_nii.get_data() == 0] = 0
+    # Save the new masked image.
     new_nii = nb.Nifti1Image(data, in_nii.affine, in_nii.header)
     new_nii.to_filename(new_name)
     return os.path.abspath(new_name)
 
+
+def create_cfm(in_file, out_path, lesion_mask, global_mask=True):
+    """
+    Create a mask to constrain registration.
+
+    Parameters
+    ----------
+    in_file : str
+        Path to an existing image (usually a mask).
+        If global_mask = True, this is used as a size/dimension reference.
+    out_path : str
+        Path/filename for the new cost function mask.
+    lesion_mask : str, optional
+        Path to an existing binary lesion mask.
+    global_mask : bool
+        Create a whole-image mask (True) or limit to reference mask (False)
+
+    Returns
+    -------
+    str
+        Absolute path of the new cost function mask.
+
+    Notes
+    -----
+    in_file and lesion_mask must be in the same
+    image space and have the same dimensions
+    """
+    import os
+    import numpy as np
+    import nibabel as nb
+    
+    # Load the input image
+    in_img = nb.load(in_file)
+    in_data = in_img.get_data()
+
+    # If we want a global mask, create one based on the input image.
+    if global_mask is True:
+        # Create a mask of ones with the shape of the input image.
+        in_data = np.ones_like(in_data, dtype=np.uint8)
+
+    # If a lesion mask was provided, combine it with the secondary mask.
+    if lesion_mask is not None:
+        # Reorient the lesion mask and get the data.
+        lm_img = nb.as_closest_canonical(nb.load(lesion_mask))
+        lm_data = lm_img.get_data()
+
+        # Subtract the lesion mask from the secondary mask.
+        cfm_data = in_data - lm_data
+        cfm_data[cfm_data < 0] = 0
+
+        # Create the cost function mask image from the subtraction.
+        cfm_img = nb.Nifti1Image(cfm_data, in_img.affine, in_img.header)
+    else:
+        # Confirm that global masking is enabled.
+        assert (global_mask is True), "If no lesion mask is provided, global_mask must be True"
+
+        # Create the cost function mask from the global mask.
+        cfm_img = nb.Nifti1Image(in_data, in_img.affine, in_img.header)
```

Comment:
This can be done for all execution paths (see my PR), since you are checking all possible options. That would also make the logic easier if you also check for the correctness the way I suggested (because that `else:` case would disappear).

(requires applying first the comment on line 428)